cot/src/config.rs

```diff
@@ -984,24 +1224,150 @@ impl From<&str> for DatabaseUrl {
-        let mut new_url = self.0.clone();
-        if !new_url.username().is_empty() {
-            new_url
-                .set_username("********")
-                .expect("set_username should succeed if username is present");
-        }
-        if new_url.password().is_some() {
-            new_url
-                .set_password(Some("********"))
-                .expect("set_password should succeed if password is present");
-        }
+        let new_url = conceal_url_parts(&self.0);
 
         f.debug_tuple("DatabaseUrl")
             .field(&new_url.as_str())
             .finish()
     }
 }
 
+/// An error returned when parsing a `CacheType` from a string.
+#[derive(Debug, Error, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ParseCacheTypeError {
+    /// The input did not match any supported cache type.
+    #[error("unsupported cache type: `{0}`")]
+    Unsupported(String),
+}
+
+/// A structure that holds the type of Cache.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[cfg(feature = "cache")]
+#[non_exhaustive]
+pub enum CacheType {
+    /// A redis cache type.
+    #[cfg(feature = "redis")]
+    Redis,
+}
+
+#[cfg(feature = "cache")]
+impl TryFrom<&str> for CacheType {
+    type Error = ParseCacheTypeError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        match value {
+            #[cfg(feature = "redis")]
+            "redis" => Ok(CacheType::Redis),
+            other => Err(ParseCacheTypeError::Unsupported(other.to_owned())),
+        }
+    }
+}
+
+#[cfg(feature = "cache")]
+impl std::str::FromStr for CacheType {
+    type Err = ParseCacheTypeError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        CacheType::try_from(s)
+    }
+}
+
+/// A URL for caches.
+///
+/// This is a wrapper over the [`url::Url`] type, which is used to store the
+/// URL of a cache. It parses the URL and ensures that it is valid.
+///
+/// # Security
+///
+/// The implementation of the [`Debug`] trait for this type hides the password
+/// from the debug output.
+///
+/// # Examples
+///
+/// ```
+/// use cot::config::CacheUrl;
+///
+/// let url = CacheUrl::from("redis://user:password@localhost:6379/0");
+/// ```
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[serde(transparent)]
+#[cfg(feature = "cache")]
+pub struct CacheUrl(url::Url);
+
+#[cfg(feature = "cache")]
+impl CacheUrl {
+    /// Returns the string representation of the cache URL.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use cot::config::CacheUrl;
+    ///
+    /// let url = CacheUrl::from("redis://user:password@localhost:6379/0");
+    /// assert_eq!(url.as_str(), "redis://user:password@localhost:6379/0");
+    /// ```
+    #[must_use]
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+#[cfg(feature = "cache")]
+impl From<String> for CacheUrl {
+    fn from(url: String) -> Self {
+        Self(url::Url::parse(&url).expect("invalid  cache URL"))
```

Comment:
@m4tx I drew inspiration from the `DbUrl ` implementation here. However, if this is Fallible, shouldn't we be implementing the `TryFrom` trait instead?

Reply:
I believe we panic here instead of handling the error on a principle that if a configuration is incorrect, the application should not start anyway.